### PR TITLE
Use Default Critical (Non-Critical) Marks for CHIP Cert Extensions.

### PR DESCRIPTION
**Pull request by emargolis**
The following certificate extensions, when present, MUST be marked as critical:
  - basic constraints extension
  - key usage extension
  - extended key usage extension

The following certificate extensions, when present, MUST be marked as non-critical:
  - subject key identifier extension
  - authority key identifier extension

These rules should always be enforced when processing X.509 representation of the certificate.
In the CHIP certificate representation the "critical" field is omitted.

Ticket: #4708

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/credentials/CHIPCertFromX509.cpp [hunk 1]
```diff
@@ -349,49 +306,53 @@ static CHIP_ERROR ConvertExtension(ASN1Reader & reader, TLVWriter & writer)
         {
             if (extensionOID == kOID_Extension_AuthorityKeyIdentifier)
             {
-                err = writer.StartContainer(ContextTag(kTag_AuthorityKeyIdentifier), kTLVType_Structure, outerContainer);
-                SuccessOrExit(err);
+                // This extention MUST be marked as non-critical.
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertFromX509.cpp:309]:
```suggestion
                // This extension MUST be marked as non-critical.
```

File: src/credentials/CHIPCertFromX509.cpp [hunk 2]
```diff
@@ -349,49 +306,53 @@ static CHIP_ERROR ConvertExtension(ASN1Reader & reader, TLVWriter & writer)
...
-                err = ConvertAuthorityKeyIdentifierExtension(reader, writer);
-                SuccessOrExit(err);
+                    err = reader.Next();
+                    VerifyOrExit(err == ASN1_END, err = ASN1_ERROR_INVALID_ENCODING);
+                }
+                ASN1_EXIT_SEQUENCE;
             }
             else if (extensionOID == kOID_Extension_SubjectKeyIdentifier)
             {
+                // This extention MUST be marked as non-critical.
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertFromX509.cpp:336]:
```suggestion
                // This extension MUST be marked as non-critical.
```

File: src/credentials/CHIPCertFromX509.cpp [hunk 3]
```diff
@@ -349,49 +306,53 @@ static CHIP_ERROR ConvertExtension(ASN1Reader & reader, TLVWriter & writer)
...
-                }
+                VerifyOrExit(reader.GetValueLen() == kKeyIdentifierLength, err = ASN1_ERROR_INVALID_ENCODING);
 
-                err = writer.PutBytes(ContextTag(kTag_SubjectKeyIdentifier_KeyIdentifier), reader.GetValue(), reader.GetValueLen());
+                err = writer.PutBytes(ContextTag(kTag_SubjectKeyIdentifier), reader.GetValue(), reader.GetValueLen());
                 SuccessOrExit(err);
             }
             else if (extensionOID == kOID_Extension_KeyUsage)
             {
+                // This extention MUST be marked as critical.
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertFromX509.cpp:350]:
```suggestion
                // This extension MUST be marked as critical.
```

File: src/credentials/CHIPCertFromX509.cpp [hunk 4]
```diff
@@ -406,11 +367,14 @@ static CHIP_ERROR ConvertExtension(ASN1Reader & reader, TLVWriter & writer)
                                  KeyUsageFlags::kCRLSign, KeyUsageFlags::kEncipherOnly, KeyUsageFlags::kEncipherOnly),
                              err = ASN1_ERROR_INVALID_ENCODING);
 
-                err = writer.Put(ContextTag(kTag_KeyUsage_KeyUsage), keyUsageBits);
+                err = writer.Put(ContextTag(kTag_KeyUsage), keyUsageBits);
                 SuccessOrExit(err);
             }
             else if (extensionOID == kOID_Extension_BasicConstraints)
             {
+                // This extention MUST be marked as critical.
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertFromX509.cpp:375]:
```suggestion
                // This extension MUST be marked as critical.
```

File: src/credentials/CHIPCertFromX509.cpp [hunk 5]
```diff
@@ -437,44 +401,36 @@ static CHIP_ERROR ConvertExtension(ASN1Reader & reader, TLVWriter & writer)
...
+                    err = writer.EndContainer(outerContainer);
+                    SuccessOrExit(err);
                 }
                 ASN1_EXIT_SEQUENCE;
             }
             else if (extensionOID == kOID_Extension_ExtendedKeyUsage)
             {
-                err = writer.StartContainer(ContextTag(kTag_ExtendedKeyUsage), kTLVType_Structure, outerContainer);
-                SuccessOrExit(err);
+                // This extention MUST be marked as critical.
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertFromX509.cpp:430]:
```suggestion
                // This extension MUST be marked as critical.
```

File: src/credentials/CHIPCertToX509.cpp [hunk 6]
```diff
@@ -310,34 +310,31 @@ static CHIP_ERROR DecodeConvertSubjectPublicKeyInfo(TLVReader & reader, ASN1Writ
 static CHIP_ERROR DecodeConvertAuthorityKeyIdentifierExtension(TLVReader & reader, ASN1Writer & writer,
                                                                ChipCertificateData & certData)
 {
-    CHIP_ERROR err, nextRes;
+    CHIP_ERROR err;
     uint32_t len;
 
     certData.mCertFlags.Set(CertFlags::kExtPresent_AuthKeyId);
 
+    // AuthorityKeyIdentifier extention MUST be marked as non-critical (default).
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertToX509.cpp:318]:
```suggestion
    // AuthorityKeyIdentifier extension MUST be marked as non-critical (default).
```

File: src/credentials/CHIPCertToX509.cpp [hunk 7]
```diff
@@ -353,13 +350,15 @@ static CHIP_ERROR DecodeConvertSubjectKeyIdentifierExtension(TLVReader & reader,
 
     certData.mCertFlags.Set(CertFlags::kExtPresent_SubjectKeyId);
 
+    // SubjectKeyIdentifier extention MUST be marked as non-critical (default).
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertToX509.cpp:353]:
```suggestion
    // SubjectKeyIdentifier extension MUST be marked as non-critical (default).
```

File: src/credentials/CHIPCertToX509.cpp [hunk 8]
```diff
@@ -529,30 +549,17 @@ static CHIP_ERROR DecodeConvertExtension(TLVReader & reader, ASN1Writer & writer
...
     ASN1_START_SEQUENCE
     {
         // extnID OBJECT IDENTIFIER,
         ASN1_ENCODE_OBJECT_ID(extensionOID);
 
-        // critical BOOLEAN DEFAULT FALSE,
-        nextRes = reader.Next();
-        VerifyOrExit(nextRes == CHIP_NO_ERROR || nextRes == CHIP_END_OF_TLV, err = nextRes);
-        if (reader.GetTag() == ContextTag(kTag_BasicConstraints_Critical))
+        // BasicConstraints, KeyUsage and ExtKeyUsage extentions MUST be marked as critical.
```
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertToX509.cpp:558]:
```suggestion
        // BasicConstraints, KeyUsage and ExtKeyUsage extensions MUST be marked as critical.
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Comment by woody-apple** [comment]
@saurabhst ?

**Review by woody-apple — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Pull request merged**